### PR TITLE
chore: add missing validations in handle_public_randomness_commit

### PR DESCRIPTION
**Pull request by gusin13**
Fixes https://github.com/babylonlabs-io/rollup-bsn-contracts/issues/46
and some other missing validations.

I compared the go code in Babylon and rust implementation, after this PR we have all the validations.

Validation Check | Go Implementation | Rust Implementation | Status
-- | -- | -- | --
Empty FP BTC PubKey | ✅ ValidateBasic() | ✅ validate_pub_rand_commit() | ✅ Complete
32-byte commitment | ✅ ValidateBasic() | ✅ validate_pub_rand_commit() | ✅ Complete
Integer overflow protection | ✅ ValidateBasic() | ✅ validate_pub_rand_commit() | ✅ Complete
Empty signature | ✅ ValidateBasic() | ✅ validate_pub_rand_commit() | ✅ Complete
Future height limit | ✅ CommitPubRandList() | ✅ validate_pub_rand_commit() | ✅ Complete
Min pub rand requirement | ✅ CommitPubRandList() | ✅ validate_pub_rand_commit() | ✅ Complete
FP exists validation | ✅ CommitPubRandList() | ✅ handle_public_randomness_commit() | ✅ Complete
Signature verification | ✅ CommitPubRandList() | ✅ handle_public_randomness_commit() | ✅ Complete
Height overlap validation | ✅ CommitPubRandList() | ✅ insert_pub_rand_commit() | ✅ Complete
Signer address validation | ✅ ValidateBasic() | ❌ (CosmWasm handles this) | ✅ Not needed

Note: Konrad fixed the 32-byte commitment check in [this PR](https://github.com/babylonlabs-io/rollup-bsn-contracts/pull/58). Once it's merged, I'll pull the changes and remove my related change.

### CODE REVIEW

File: contracts/finality/src/contract.rs [hunk 1]
```diff
@@ -7,23 +7,27 @@ use crate::queries::query_block_voters;
...
+use cosmwasm_std::{to_json_binary, Deps, DepsMut, Env, MessageInfo, QueryResponse, Response};
 use cw_controllers::AdminError;
 
 pub fn instantiate(
     mut deps: DepsMut<BabylonQuery>,
     _env: Env,
     _info: MessageInfo,
     msg: InstantiateMsg,
-) -> StdResult<Response<BabylonMsg>> {
+) -> Result<Response<BabylonMsg>, ContractError> {
```
Comment (SebastianElvis) [inline contracts/finality/src/contract.rs:18]: L18 does not need to be changed imo

File: contracts/finality/src/contract.rs [hunk 2]
```diff
@@ -146,6 +150,7 @@ pub(crate) mod tests {
             admin: init_admin.to_string(),
             consumer_id,
             is_enabled: true,
+            min_pub_rand: 1,
```
Comment (SebastianElvis) [inline contracts/finality/src/contract.rs:153]: For tests I think we can pick a bigger value. There is already validation that pub rand commit should contain >=1 pub rand values

File: contracts/finality/src/exec/finality.rs [hunk 3]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
...
+    }
+
+    // Check if start height is too far into the future
+    let current_height = env.block.height;
+    if start_height >= current_height + MAX_PUB_RAND_COMMIT_OFFSET {
+        return Err(ContractError::InvalidPubRandHeight(
+            start_height,
+            current_height + MAX_PUB_RAND_COMMIT_OFFSET - 1,
+        ));
+    }
```
Comment (SebastianElvis) [inline contracts/finality/src/exec/finality.rs:125]: I'm not aware if Babylon has this. If Babylon does not have this then we don't need to have it in the contract either.

In addition, this validation does not achieve the supposed goal -- `env.block.height` is the height of *Babylon* block, but the pub rand commit is associated with rollup BSN blocks

File: contracts/finality/src/exec/finality.rs [hunk 4]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
     Ok(Response::new().add_event(event))
 }
 
-// Copied from contracts/btc-staking/src/finality.rs
+/// Validates all parameters for public randomness commitment
+fn validate_pub_rand_commit(
+    env: &Env,
```
Comment (SebastianElvis) [inline contracts/finality/src/exec/finality.rs:82]: Here `env` shall not be used for the validation. The `env` contains information about Babylon chain, but not the rollup BSN. The checks should not be against the Babylon chain

File: contracts/finality/src/exec/finality.rs [hunk 5]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
...
+        return Err(ContractError::EmptyFpBtcPubKey);
+    }
+
+    // Check if commitment is exactly 32 bytes
+    if commitment.len() != EXPECTED_COMMITMENT_LENGTH_BYTES {
+        return Err(ContractError::InvalidCommitmentLength {
+            expected: EXPECTED_COMMITMENT_LENGTH_BYTES,
+            actual: commitment.len(),
+        });
+    }
```
Comment (SebastianElvis) [inline contracts/finality/src/exec/finality.rs:101]: https://github.com/babylonlabs-io/rollup-bsn-contracts/pull/58 will be subsumed by this PR cc @KonradStaniec?

File: contracts/finality/src/exec/finality.rs [hunk 6]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
...
+        return Err(ContractError::OverflowInBlockHeight {
+            start_height,
+            end_height,
+        });
+    }
+
+    // Check if signature is empty
+    if signature.is_empty() {
+        return Err(ContractError::EmptySignature);
+    }
```
Comment (SebastianElvis) [inline contracts/finality/src/exec/finality.rs:116]: Signature also needs to be 32 bytes. Perhaps the `EXPECTED_COMMITMENT_LENGTH_BYTES` can be renamed to sth like `HASH_SIZE` and is used for checking the length of many variables

File: contracts/finality/src/exec/finality.rs (hunk 4 above)
Comment (SebastianElvis) [reply contracts/finality/src/exec/finality.rs:82]: In line with the above comment, this function will then become a sanity check function for pub rand commit requests. I think we can put it as a function under `ExecuteMsg::CommitPublicRandomness`

File: contracts/finality/src/exec/finality.rs [hunk 7]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
...
+        ));
+    }
+
+    // Validate minimum public randomness requirement
+    if num_pub_rand < min_pub_rand {
+        return Err(ContractError::TooFewPubRand {
+            given: num_pub_rand,
+            required: min_pub_rand,
+        });
+    }
```
Comment (SebastianElvis) [inline contracts/finality/src/exec/finality.rs:133]: https://github.com/babylonlabs-io/rollup-bsn-contracts/blob/5fe461327296d0cbcfc9a3ea9b8a245ad28e44b4/contracts/finality/src/state/public_randomness.rs#L169-L172 can be removed due to this change

File: contracts/finality/src/exec/finality.rs (hunk 5 above)
Comment (KonradStaniec) [reply contracts/finality/src/exec/finality.rs:101]: yep :) will close the old one

**Review by vitsalis — DISMISSED** [review]
Nice work. Some comments

File: contracts/finality/src/exec/finality.rs (hunk 4 above)
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs:82]: ok 
- i have removed the env check, that means we won't be able to verify future offset validation
- have ported the code to new file



> I think we can put it as a function under ExecuteMsg::CommitPublicRandomness

> how about we create a new exec

can you elaborate this again, i didn't get properly. 
what does creating a new exec fn mean here? why do we need to create a new one?
there would still be a single exec fn `ExecuteMsg::CommitPublicRandomness` just the handler would live in a new file no?

File: contracts/finality/src/exec/finality.rs [hunk 8]
```diff
@@ -18,13 +20,25 @@ use k256::sha2::{Digest, Sha256};
...
+    let config = get_config(deps.as_ref())?;
+    validate_pub_rand_commit(
+        env,
+        fp_btc_pk_hex,
+        start_height,
+        num_pub_rand,
+        commitment,
+        signature,
+        config.min_pub_rand,
+    )?;
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs:40]: can you elaborate?

we have these checks in babylon as well
[checking if fp btc pk is not nil](https://github.com/babylonlabs-io/babylon/blob/12953b4b08966867f1a8e92c0e92c37b682d0d77/x/finality/types/msg.go#L45)
[checking if sig empty](https://github.com/babylonlabs-io/babylon/blob/12953b4b08966867f1a8e92c0e92c37b682d0d77/x/finality/types/msg.go#L65)

don't we need these checks?

File: contracts/finality/src/exec/finality.rs [hunk 9]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
...
+    env: &Env,
+    fp_btc_pk_hex: &str,
+    start_height: u64,
+    num_pub_rand: u64,
+    commitment: &[u8],
+    signature: &[u8],
+    min_pub_rand: u64,
+) -> Result<(), ContractError> {
+    // Check if FP BTC PubKey is empty
+    if fp_btc_pk_hex.is_empty() {
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs:91]: these checks are similar to validateBasic() in babylon code, are you suggesting to rename the fn to be more generic instead of `validate_pub_rand_commit` if this is confusing?

File: contracts/finality/src/exec/finality.rs (hunk 3 above)
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: this check does exist in babylon [here](https://github.com/babylonlabs-io/babylon/blob/12953b4b08966867f1a8e92c0e92c37b682d0d77/x/finality/keeper/msg_server.go#L267)

but i agree in rollup contract we can't verify this as we don't have access to rollup bsn height, have removed this check now

File: contracts/finality/src/exec/finality.rs [hunk 10]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
...
+        });
+    }
+
+    // Check if signature is empty
+    if signature.is_empty() {
+        return Err(ContractError::EmptySignature);
+    }
+
+    // Check if start height is too far into the future
+    let current_height = env.block.height;
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: we removed this env thing anyways as the contract doesn't know rollup bsn height, the env.block.height is basically babylon height which should not be used for validations

File: contracts/finality/src/exec/finality.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: hmm, i have moved the fp btc pk hex check outside the `validate_pub_rand_commit` fn now

File: contracts/finality/src/exec/finality.rs [hunk 11]
```diff
@@ -63,7 +77,64 @@ pub fn handle_public_randomness_commit(
...
+    let end_height = start_height.saturating_add(num_pub_rand);
+    if start_height >= end_height {
+        return Err(ContractError::OverflowInBlockHeight {
+            start_height,
+            end_height,
+        });
+    }
+
+    // Check if signature is empty
+    if signature.is_empty() {
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: hmm, i have moved the signature check outside the `validate_pub_rand_commit` fn now

File: contracts/finality/src/exec/finality.rs (hunk 6 above)
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: hm i see, do we have such check in babylon? i couldn't find

File: contracts/finality/src/exec/finality.rs [hunk 12]
```diff
@@ -450,4 +521,217 @@ pub(crate) mod tests {
...
     }
+
+    mod validation_tests {
+        use super::*;
+        use crate::contract::tests::mock_deps_babylon;
+        use crate::state::config::{Config, CONFIG};
+        use cosmwasm_std::testing::mock_env;
+
+        #[test]
+        fn test_empty_fp_btc_pk_fails() {
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: but we passing the `fp_btc_pk` to `handle_public_randomness_commit` and we need to test if this empty then whats the behaviour?

can you elaborate the ask?

File: contracts/finality/src/exec/finality.rs [hunk 13]
```diff
@@ -450,4 +521,217 @@ pub(crate) mod tests {
...
+                result.unwrap_err(),
+                ContractError::InvalidCommitmentLength {
+                    expected: EXPECTED_COMMITMENT_LENGTH_BYTES,
+                    actual: EXPECTED_COMMITMENT_LENGTH_BYTES + 1
+                }
+            );
+        }
+
+        #[test]
+        fn test_empty_signature_fails() {
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: i am bit confused, the sig check still exists in `handle_public_randomness_commit` so need to test it no?

the suggestion in prev comments is to move the sig check outside `validate_pub_rand_commit` which is fixed now but the sig check will still exists in the handler and this test is testing the handler itself.

File: contracts/finality/src/exec/finality.rs (hunk 6 above)
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: bit confused infact how come 32 bytes? in babylon the bip340 sig size is 64 bytes [ref](https://github.com/babylonlabs-io/babylon/blob/12953b4b08966867f1a8e92c0e92c37b682d0d77/types/btc_schnorr_sig.go#L12)

File: contracts/finality/src/exec/finality.rs [hunk 14]
```diff
@@ -450,4 +521,217 @@ pub(crate) mod tests {
...
+                100,
+                &[0u8; EXPECTED_COMMITMENT_LENGTH_BYTES],
+                &[1u8; 64],
+            );
+
+            assert_eq!(result.unwrap_err(), ContractError::EmptyFpBtcPubKey);
+        }
+
+        #[test]
+        fn test_invalid_commitment_length_fails() {
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: fixed, using datagen methods now

File: contracts/finality/src/contract.rs [hunk 15]
```diff
@@ -211,4 +190,66 @@ pub(crate) mod tests {
...
     }
+
+    mod property_tests {
+        use super::*;
+        use proptest::prelude::*;
+
+        proptest! {
+            #[test]
+            fn test_instantiate_validation(
+                min_pub_rand in 0u64..1000000,
```
Comment (gusin13) [inline contracts/finality/src/contract.rs:201]: i could use datagen method, will remove this test and use datagen to be consistent

File: contracts/finality/src/exec/finality.rs (hunk 6 above)
Comment (SebastianElvis) [reply contracts/finality/src/exec/finality.rs]: EOTS sig and BIP340 sig is different.

In particular, BIP340 signature is a Schnorr signature encoded from `(R, s)`. Each element has 32 bytes.

EOTS sig is only the `s` part of the BIP-340 signature. `R` is the corresponding public randomness

File: contracts/finality/Cargo.toml [hunk 16]
```diff
@@ -40,6 +40,7 @@ eots             = { workspace = true }
 anyhow                = { workspace = true }
 derivative            = { workspace = true }
 rand                  = { workspace = true }
+proptest              = { workspace = true }
```
Comment (SebastianElvis) [inline contracts/finality/Cargo.toml:43]: I'm not sure if this library is common practice for writing Rust tests. Since there is only 1 unit test using this lightly, I recommend we avoid it. Wdyt?

File: contracts/finality/src/exec/public_randomness.rs [hunk 17]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+pub fn validate_pub_rand_commit(
+    start_height: u64,
+    num_pub_rand: u64,
+    commitment: &[u8],
+    min_pub_rand: u64,
+) -> Result<(), ContractError> {
+    // Check if commitment is exactly 32 bytes
+    if commitment.len() != EXPECTED_COMMITMENT_LENGTH_BYTES {
```
Comment (SebastianElvis) [inline contracts/finality/src/exec/public_randomness.rs:83]:
```suggestion
    if commitment.len() != COMMITMENT_LENGTH_BYTES {
```

**Review by SebastianElvis — APPROVED** [review]
LGTM with minor comments. Thanks!

File: contracts/finality/src/contract.rs (hunk 1 above)
Comment (SebastianElvis) [reply contracts/finality/src/contract.rs:19]: Hmm checked other endpoint functions and all of them return `Result<Response<BabylonMsg>, ContractError>`. So all good, sorry for the confusion

File: contracts/finality/Cargo.toml (hunk 16 above)
Comment (gusin13) [reply contracts/finality/Cargo.toml:43]: yes, agreed I'll remove this soon. I added this fuzz style test which imported this library, I think its better to add methods in datagen and use unit tests.

File: contracts/finality/src/exec/finality.rs (hunk 6 above)
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: I see, sorry I am still bit confused. 

The local deployment [code ref ](https://github.com/babylonlabs-io/babylon-bsn-integration-deployment/blob/938b9d0159bd978634ef860ed767f994c0040c19/deployments/rollup-bsn-demo/crypto-ops-tool/cmd/crypto-ops/main.go#L541) is generating BIP340 sig and then sending this to contract [code ref](https://github.com/babylonlabs-io/babylon-bsn-integration-deployment/blob/938b9d0159bd978634ef860ed767f994c0040c19/deployments/rollup-bsn-demo/demo.sh#L260-L296), so the sig we receive in contract should be a serialized Schnorr sig (64 bytes) no?

File: contracts/finality/src/contract.rs [hunk 18]
```diff
@@ -7,23 +7,27 @@ use crate::queries::query_block_voters;
...
+        return Err(ContractError::InvalidMinPubRand(msg.min_pub_rand));
+    }
+
     let api = deps.api;
     ADMIN.set(deps.branch(), Some(api.addr_validate(&msg.admin)?))?;
     IS_ENABLED.save(deps.storage, &msg.is_enabled)?;
 
     let config = Config {
         consumer_id: msg.consumer_id,
+        min_pub_rand: msg.min_pub_rand,
```
[the remark this answers is not in the document]
Comment (vitsalis) [reply contracts/finality/src/contract.rs:31]: In the sense that this is an additional configuration of the module, while the rest of the PR is about additional validations. eitherway, np.

File: contracts/finality/src/exec/finality.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (vitsalis) [reply contracts/finality/src/exec/finality.rs]: Ah you're right, I thought this function was testing `validate_pub_rand_commit`.

File: contracts/finality/src/exec/finality.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (vitsalis) [reply contracts/finality/src/exec/finality.rs]: Ah you're right, I thought this function was testing validate_pub_rand_commit.

File: contracts/finality/src/exec/finality.rs (hunk 4 above)
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs]: hmm i asked ai, this is basically a code style change and would need bunch of stuff

first we need new struct
```
#[derive(Debug, Clone)]
pub struct CommitPublicRandomnessData {
    pub fp_pubkey_hex: String,
    pub start_height: u64,
    pub num_pub_rand: u64,
    pub commitment: Vec<u8>,  // <- Raw bytes for internal use
    pub signature: Vec<u8>,   // <- Raw bytes for internal use
}

impl CommitPublicRandomnessData {
    pub fn validate(&self, min_pub_rand: u64) -> Result<(), ContractError> {
        // validation logic here
    }
}
```

then we need to fill data in this struct on exec msg
```
ExecuteMsg::CommitPublicRandomness {
    fp_pubkey_hex,
    start_height,
    num_pub_rand,
    commitment,
    signature,
} => {
    // Convert from message format to business logic format
    let commit_data = CommitPublicRandomnessData {
        fp_pubkey_hex,
        start_height,
        num_pub_rand,
        commitment: commitment.to_vec(),  // Binary -> Vec<u8>
        signature: signature.to_vec(),    // Binary -> Vec<u8>
    };
    
    handle_public_randomness_commit(deps, &env, commit_data)
}
```

finally the handler might look like this then

```
pub fn handle_public_randomness_commit(
    deps: DepsMut<BabylonQuery>,
    _env: &Env,
    commit_data: CommitPublicRandomness,  // <- Now takes struct instead of individual params
) -> Result<Response<BabylonMsg>, ContractError> {
    let config = get_config(deps.as_ref())?;
    
    // Validate the struct (this replaces the current validation logic)
    commit_data.validate(config.min_pub_rand)?;

    // All the rest stays the same:
    ensure_fp_exists_and_not_slashed(deps.as_ref(), &commit_data.fp_pubkey_hex)?;
    
    let fp_btc_pk = hex::decode(&commit_data.fp_pubkey_hex)?;
    
    verify_commitment_signature(
        &fp_btc_pk,
        commit_data.start_height,
        commit_data.num_pub_rand,
        &commit_data.commitment,
        &commit_data.signature,
    )?;

    let current_epoch = get_current_epoch(&deps.as_ref())?;
    insert_pub_rand_commit(/* ... */)?;

    // Build and return response
    Ok(Response::new().add_event(event))
}
```

i'd prefer to do in another pr as this code style change, we'd need to adapt this paradigm then in all handlers (same for cosmos contracts as well) 
which is a broad change, created ticket for this
https://github.com/babylonlabs-io/rollup-bsn-contracts/issues/67

Comment (SebastianElvis) [reply contracts/finality/src/exec/finality.rs]: Hmm I have given it a try and this refactoring does not seem very beneficial. In particular many sanity checks are actually covered by later stateful checks. I think the current implementation is good enough and maybe we don't need this refactoring. Wdyt?
